src/passes/ConstantFieldPropagation.cpp

```diff
@@ -0,0 +1,393 @@
+    }
+
+    // If the value is not a constant, then it is unknown and we must give up.
+    if (!info.isConstant()) {
+      return;
+    }
+
+    // We can do this! Replace the get with a throw on a null reference (as the
+    // get would have done so), plus the constant value. (Leave it to further
+    // optimizations to get rid of the ref.)
+    replaceCurrent(builder.makeSequence(
+      builder.makeDrop(builder.makeRefAs(RefAsNonNull, curr->ref)),
+      builder.makeConstantExpression(info.getConstantValue())));
+    changedTypes = true;
+  }
+
+  void doWalkFunction(Function* func) {
+    WalkerPass<PostWalker<FunctionOptimizer>>::doWalkFunction(func);
+
+    // If we changed any types, we need to update parents.
+    if (changedTypes) {
+      ReFinalize().walkFunctionInModule(func, getModule());
+    }
+  }
+
+private:
+  StructValuesMap* infos;
+
+  bool changedTypes = false;
+};
+
+struct ConstantFieldPropagation : public Pass {
+  void run(PassRunner* runner, Module* module) override {
+    if (getTypeSystem() != TypeSystem::Nominal) {
+      Fatal() << "ConstantFieldPropagation requires nominal typing";
+    }
+
+    // Find and analyze all writes inside each function.
+    FunctionStructValuesMap functionInfos;
+    for (auto& func : module->functions) {
+      // Initialize the data for each function, so that we can operate on this
+      // structure in parallel without modifying it.
+      functionInfos[func.get()];
+    }
+    Scanner scanner(&functionInfos);
+    scanner.run(runner, module);
+    scanner.walkModuleCode(module);
+
+    // Combine the data from the functions.
+    StructValuesMap combinedInfos;
+    for (auto& kv : functionInfos) {
+      StructValuesMap& infos = kv.second;
+      for (auto& kv : infos) {
+        auto type = kv.first;
+        auto& info = kv.second;
+        auto& combinedInfo = combinedInfos[type];
+        for (Index i = 0; i < info.size(); i++) {
+          combinedInfo[i].combine(info[i]);
+        }
+      }
+    }
+
+    // Handle subtyping. |combinedInfo| so far contains data that represents
+    // each struct.new and struct.set's operation on the struct type used in
+    // that instruction. That is, if we do a struct.set to type T, the value was
+    // noted for type T. But our actual goal is to answer questions about
+    // struct.gets. Specifically, when later we see:
+    //
+    //  (struct.get $A x (REF-1))
+    //
+    // Then we want to find all struct.sets
+    //
+    //  (struct.set $B x (REF-2) (..value..))
+    //
+    // where $B is a subtype of $A, because at runtime the value REF-1 might not
+    // only be of type $A but also any subtype of $A. To make the lookup
+    // efficient when we see the get, we propagate information about sets to
+    // their supers, recursively, so long as the super still has that field.
+    // (Note that this also handles the case of REF-2 being a subtype of $B.)
+    //
+    // TODO: Are cycles possible with nominal supertypes?
```

Comment:
No, cycles are disallowed. Would it be worth a TODO to visit the types in topological order to avoid repeated work?

Reply:
Even this, when no field is added?
```wat
(type $A (struct i32 (extends $B)))
(type $B (struct i32 (extends $A)))
```
If so, then different nominal types can never be equivalent, IIUC?

I'll add a TODO.